### ISSUE TITLE
P1.3: Challenge Caching & Pattern Learning with ML Classifier

### ISSUE THREAD

**Issue by enuno** [enuno/moltbot-philosopher#34]
## Description

Implement graduated TTL caching for verified challenges and ML-based difficulty classification to improve cache hit rate and model selection accuracy.

## Context
- Depends on: P1.1 (preprocessing), P1.2 (model selection)
- Target: Cache hit rate >60% for repeated challenges
- Pattern classifier improves P1.2 model selection accuracy

### Unit Tests: Cache TTL Assignment

Create `services/intelligent-proxy/test/caching.test.js`:

1. **Test: High Confidence Solution (≥0.95)**
   - Cache TTL should be 7 days
   - Verify: Redis SET with EX 604800 (7 days in seconds)

2. **Test: Medium Confidence Solution (0.8-0.94)**
   - Cache TTL should be 3 days
   - Verify: Redis SET with EX 259200 (3 days in seconds)

3. **Test: Low Confidence Solution (0.6-0.79)**
   - Cache TTL should be 24 hours
   - Verify: Redis SET with EX 86400

4. **Test: Failed Solution**
   - Cache TTL should be 12 hours
   - Store reason ("model_mismatch", "timeout", "parse_error")
   - Verify: Can differentiate failure types on retry

5. **Test: Cache Hit**
   - Store challenge → solution mapping in Redis
   - Query same challenge again
   - Verify: Returns cached value without model call
   - Track hit/miss metrics

6. **Test: Cache Determinism**
   - Same challenge preprocessed twice
   - Cache key should be identical (deterministic hash)
   - Verify: Consistent cache lookups

### Unit Tests: Pattern Classification ML

Create `services/intelligent-proxy/test/pattern-classifier.test.js`:

1. **Test: Feature Extraction**
   - Input: Preprocessed challenge text
   - Extract: [symbol_density, pattern_diversity, length, vocab_rarity]
   - Verify: Features are numeric and normalized (0-1 range)

2. **Test: Classifier Training**
   - Train on 100 labeled challenges (easy/medium/hard + actual difficulty 1-10)
   - Verify: Logistic regression model created
   - Test accuracy on holdout set ≥70%

3. **Test: Difficulty Prediction**
   - Input: New challenge features
   - Output: difficulty_score (1-10) with confidence
   - Verify: Predictions align with human ratings (correlation ≥0.7)

4. **Test: Classifier Persistence**
   - Train model → serialize to disk
   - Load from disk → verify predictions identical
   - Compare with in-memory model

5. **Test: Feature Engineering**
   - Symbol density = (symbol_count / total_chars) * 100
   - Pattern diversity = unique_character_sets
   - Length = challenge text length
   - Vocab rarity = avg_character_frequency_rank
   - Verify: Features are meaningful for classification

### Integration Tests: Cache + Classifier Pipeline

Create `services/intelligent-proxy/test/caching-integration.test.js`:

1. **Test: Cache Hit Rate**
   - Process 100 challenges
   - Store in cache with TTL
   - Process same 100 challenges again
   - Verify: Cache hit rate >80% (identical challenges reuse cache)

2. **Test: Pattern Learning**
   - Train classifier on first 50 challenges
   - Use classifier to predict model for next 50
   - Verify: Model accuracy improves by ≥5% vs P1.2 baseline

3. **Test: Cache Expiration Accuracy**
   - Store challenge with 1-second TTL (for testing)
   - Verify present immediately after
   - Wait 2 seconds
   - Verify expired (returns null)

4. **Test: End-to-End: Cache → Classifier → Model**
   - Challenge comes in
   - Check cache (miss)
   - Classify difficulty using ML
   - Select model based on classification + performance history
   - Run verification
   - Cache result with appropriate TTL
   - Next identical challenge hits cache immediately

5. **Test: Statistics Endpoint**
   - Call /cache/stats
   - Verify: {hit_rate: 0.65, total_hits: 1300, total_misses: 700, avg_ttl_days: 4.2}

### Files to Create/Modify
- **New:** `services/intelligent-proxy/src/cache-manager.ts`
- **New:** `services/intelligent-proxy/src/pattern-classifier.ts`
- **New:** `services/intelligent-proxy/test/caching.test.js`
- **New:** `services/intelligent-proxy/test/pattern-classifier.test.js`
- **Modify:** `services/intelligent-proxy/index.js` (integrate cache check + classifier)

### Architecture

**CacheManager**
- Redis key: `verification:challenge:{hash}:{confidence_tier}`
- Graduated TTL based on confidence
- Stats tracking: hit_rate, total_hits, total_misses, cache_size

**PatternClassifier**
- Feature extraction from preprocessed challenge
- Logistic regression model (scikit-learn or lightweight alternative)
- Train on labeled dataset (bootstrapped from manual ratings + historical performance)
- Persist model as JSON (sklearn can export to JSON)
- Prediction: difficulty_score (1-10) + confidence (0-1)

**Integration Points**
- Cache lookup before model selection (save money + latency)
- Classifier output feeds into P1.2 model selector
- Both use preprocessed challenge from P1.1 as input

### ML Model Data Flow
```
Challenge (raw)
    ↓
P1.1: Preprocessing (normalize obfuscation)
    ↓
P1.3: Feature Extraction (extract numeric features)
    ↓
P1.3: Pattern Classifier (predict difficulty 1-10)
    ↓
P1.2: Model Selector (choose model based on difficulty)
    ↓
Execute verification
    ↓
P1.3: Cache Store (graduated TTL based on confidence)
    ↓
Cache Hit on next identical challenge
```

## Success Criteria

- [ ] Cache hit rate >60% on repeated challenges
- [ ] Classifier accuracy ≥70% on holdout test set
- [ ] Classifier predictions improve model selection by ≥5%
- [ ] Combined P1.1+P1.2+P1.3: ≥70% verification success rate (maintained)
- [ ] Cache statistics endpoint working and accurate
- [ ] All integration tests passing
- [ ] Pattern classifier model persisted and loadable

## Effort Estimate
- CacheManager: 8 hours
- PatternClassifier: 10 hours
- Feature engineering: 3 hours
- Testing + ML iteration: 4 hours
- **Total: 25 hours**

## Dependencies
- **Blocks:** P1.4 (Timeout Handling) can run in parallel
- **Blocked By:** P1.1 (preprocessing), P1.2 (model selection)

## Related Issues
- P1.1: Text Preprocessing Layer (#32)
- P1.2: Multi-Model Strategy (#33)
- P1.4: Verification Timeout Handling
